openedx_events/event_bus/__init__.py

```diff
@@ -1 +1,83 @@
 """Classes and utility functions for the event bus."""
+
+from abc import ABC, abstractmethod
+from typing import NoReturn, Optional
+
+from django.conf import settings
+from django.utils.module_loading import import_string
+
+
+class EventBus(ABC):
+    """
+    Parent class for event bus implementations.
+    """
+
+    @abstractmethod
+    def send(
+            *, signal: OpenEdxPublicSignal, topic: str, event_key_field: str, event_data: dict,
+            sync: bool = False,
+    ) -> None:
+        """
+        Send a signal event to the event bus under the specified topic.
+
+        Arguments:
+            signal: The original OpenEdxPublicSignal the event was sent to
+            topic: The event bus topic for the event
+            event_key_field: Path to the event data field to use as the event key (period-delimited
+              string naming the dictionary keys to descend)
+            event_data: The event data (kwargs) sent to the signal
+            sync: Whether to wait indefinitely for event to be received by the message bus (probably
+              only want to use this for testing)
```

Comment:
Is this something we will be using now? What does this translate to? An immediate call to `flush()`? Do we need to expose `flush`? Do we want to separate Producer and Consumer code in case either gets more complex and there is confusing overlap?

Reply:
For Kafka, a call to flush. But... maybe there should instead be a different method on `EventBus` called `shutdown` or something, which can be used for testing (and later be called by some Django shutdown hook.)

At first thought I'd prefer having this be the single entry point for both consumer and producer so we don't have to load multiple things. But hmm... for Kafka it's all the same; once you have the connection settings, I think everything else is optional. I suppose it's possible that with some other implementation you might have a well-configured producer and an ill-configured consumer, and if you really just want the producer side then you wouldn't care! So it might be worth splitting after all.

Reply:
48e3918 splits it into a producer class (providing `send`) and a consumer class (providing `consume_indefinitely`). It also removes `sync`; to provide that functionality, we could later add a synchronous `shutdown()` method, which is really what it's for (besides testing).